src/io/free_functions.rs

```diff
@@ -0,0 +1,298 @@
+    }
+}
+
+pub(crate) fn save_buffer_with_format_impl(
+    path: &Path,
+    buf: &[u8],
+    width: u32,
+    height: u32,
+    color: color::ColorType,
+    format: ImageFormat,
+) -> io::Result<()> {
+    let fout = &mut BufWriter::new(File::create(path)?);
+
+    match format {
+        #[cfg(feature = "ico")]
+        image::ImageFormat::ICO => ico::ICOEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "jpeg")]
+        image::ImageFormat::JPEG => jpeg::JPEGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "png_codec")]
+        image::ImageFormat::PNG => png::PNGEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "bmp")]
+        image::ImageFormat::BMP => bmp::BMPEncoder::new(fout).encode(buf, width, height, color),
+        #[cfg(feature = "tiff")]
+        image::ImageFormat::TIFF => tiff::TiffEncoder::new(fout)
+            .encode(buf, width, height, color)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, Box::new(e))),
+        _ => Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            &format!("Unsupported image format image/{:?}", format)[..],
+        )),
+    }
+}
+
+/// Guess format from a path.
+///
+/// Returns `PathError::NoExtension` if the path has no extension or returns a
+/// `PathError::UnknownExtension` containing the extension if  it can not be convert to a `str`.
+pub(crate) fn guess_format_from_path_impl(path: &Path) -> Result<ImageFormat, PathError> {
+    let exact_ext = path.extension();
+
+    let ext = exact_ext
+        .and_then(|s| s.to_str())
+        .map(str::to_ascii_lowercase);
+
+    let ext = ext.as_ref()
+        .map(String::as_str);
+
+    Ok(match ext {
+        Some("jpg") | Some("jpeg") => image::ImageFormat::JPEG,
+        Some("png") => image::ImageFormat::PNG,
+        Some("gif") => image::ImageFormat::GIF,
+        Some("webp") => image::ImageFormat::WEBP,
+        Some("tif") | Some("tiff") => image::ImageFormat::TIFF,
+        Some("tga") => image::ImageFormat::TGA,
+        Some("bmp") => image::ImageFormat::BMP,
+        Some("ico") => image::ImageFormat::ICO,
+        Some("hdr") => image::ImageFormat::HDR,
+        Some("pbm") | Some("pam") | Some("ppm") | Some("pgm") => image::ImageFormat::PNM,
+        // The original extension is used, instead of _format
+        _format => return match exact_ext {
+            None => Err(PathError::NoExtension),
+            Some(os) => Err(PathError::UnknownExtension(os.to_owned())),
+        },
+    })
+}
+
+static MAGIC_BYTES: [(&'static [u8], ImageFormat); 17] = [
+    (b"\x89PNG\r\n\x1a\n", ImageFormat::PNG),
+    (&[0xff, 0xd8, 0xff], ImageFormat::JPEG),
+    (b"GIF89a", ImageFormat::GIF),
+    (b"GIF87a", ImageFormat::GIF),
+    (b"RIFF", ImageFormat::WEBP), // TODO: better magic byte detection, see https://github.com/image-rs/image/issues/660
+    (b"MM\x00*", ImageFormat::TIFF),
+    (b"II*\x00", ImageFormat::TIFF),
```

Comment:
Since this likely replaces #933, should remember to cherry pick the fix to magic bytes here.
@martinlindhe

Reply:
Would you say this PR is a fitting replacement?

Reply:
@HeroicKatora yea, sure lets drop #933 for this PR